### PR TITLE
00554: Rebuilds required in-memory data structures after reconnect

### LINKED ISSUE #554: Reconnect functionality in Hedera
**Summary**
Implement reconnect functionality in Hedera
Platform notifies services of a successful Reconnect 

Hedera must rebuild in-memory data structures that are not  of state (also cached in memory) from latest state before accepting and handling transactions. These include

1. Address Books // not needed this moment because it is not used
2. Dedup & Expiration data structures
3. Fees 
4. Exchange Rates
5. Application properties
6. API permissions
7. Record Stream Hash: https://github.com/hashgraph/hedera-services/issues/608
8. Split `init` into `init` and `genesisInit`  https://github.com/hashgraph/hedera-services/issues/649
9. Implement a validator for 2-6: https://github.com/hashgraph/hedera-services/issues/650

### LINKED ISSUE #650: Validate reconnect for Services
**Summary**
As a Hedera Services client, I need to implement a reconnect test for regression,  so that I can validate `hedera-node` reconnected successfully.

**Possible resolution**
Implement `ServicesReconnectValidator` which validates that the in-memory metadata from `ServicesState` is available after reconnecting. We need to validate

1. Dedup & Expiration data structures
2. Fees
3. Exchange Rates
4. Application properties
5. API permissions

### PR DESCRIPTION
**Pull request by cesaragv**
**Related issue(s)**:
Closes #554 
Closes #650 

**Summary of the change**:
Rebuilds and initializes every in-memory data structure after a reconnect happens

### CODE REVIEW

File: hedera-node/src/test/java/com/hedera/services/ServicesStateTest.java [hunk 1]
```diff
@@ -257,14 +272,25 @@ public void initsAsExpected() {
...
+
+		// then:
 		inOrder.verify(ctx).nodeAccount();
 		inOrder.verify(ctx).update(subject);
+		inOrder.verify(ctx).rebuildBackingStoresIfPresent();
+		inOrder.verify(historian).reviewExistingRecords();
+		inOrder.verify(systemFilesManager).loadApplicationProperties();
+		inOrder.verify(systemFilesManager).loadApiPermissions();
+		// TODO: uncomment below line.
+		// inOrder.verify(systemFilesManager).loadFeeSchedules();
```
Comment (QianSwirlds) [inline hedera-node/src/test/java/com/hedera/services/ServicesStateTest.java:293]: this line could be uncommented, right?

Comment (cesaragv) [reply hedera-node/src/test/java/com/hedera/services/ServicesStateTest.java]: @QianSwirlds you are right. I just uncommented it.

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 2]
```diff
@@ -1425,6 +1434,7 @@ public SystemFilesManager systemFilesManager() {
 							midnightRates().replaceWith(rates);
 						}
 					},
+					schedules -> fees().init(),
```
Comment (qnswirlds) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:1437]: Should be tested

File: hedera-node/src/main/java/com/hedera/services/fees/calculation/AwareFcfsUsagePrices.java [hunk 3]
```diff
@@ -114,12 +116,12 @@ public FeeData pricesGiven(HederaFunctionality function, Timestamp at) {
...
-		} catch (Exception ignore) {
+		} catch (Exception e) {
 			log.warn(
-					"Only default usage prices available for function {} @ {}.{}!",
+					"Only default usage prices available for function {} @ {}!",
 					function,
-					at.getSeconds(),
-					at.getNanos());
+					Instant.ofEpochSecond(at.getSeconds(), at.getNanos()),
+					e);
```
Comment (qnswirlds) [inline hedera-node/src/main/java/com/hedera/services/fees/calculation/AwareFcfsUsagePrices.java:124]: Seems missing one pair of `{}` in `log.warn()`. Should be better if the `log.warn()` is tested.

File: hedera-node/src/test/java/com/hedera/services/state/expiry/ExpiryManagerTest.java [hunk 4]
```diff
@@ -113,15 +113,37 @@ private AccountID asAccount(long num) {
...
+		// when:
+		subject.restartTrackingFrom(accounts);
+
+		// then:
+		verify(recordCache).reset();
+		verify(txnHistories).clear();
+	}
+
 	@Test
 	public void resumesTrackingAsExpected() {
```
Comment (qnswirlds) [inline hedera-node/src/test/java/com/hedera/services/state/expiry/ExpiryManagerTest.java:141]: Should be renamed into `restartsTrackingAsExpected`

File: hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryManager.java [hunk 5]
```diff
@@ -56,7 +56,11 @@ public void trackRecord(AccountID owner, long expiry) {
 		payerExpiries.track(owner.getAccountNum(), expiry);
 	}
 
-	public void resumeTrackingFrom(FCMap<MerkleEntityId, MerkleAccount> accounts) {
+	public void restartTrackingFrom(FCMap<MerkleEntityId, MerkleAccount> accounts) {
+		recordCache.reset();
+		txnHistories.clear();
+		payerExpiries.reset();
```
Comment (qnswirlds) [inline hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryManager.java:62]: The call to `payerExpiries.reset()` was not tested.

**Review by qnswirlds — CHANGES_REQUESTED** [review]

File: hedera-node/src/test/java/com/hedera/services/state/expiry/ExpiryManagerTest.java (hunk 4 above)
Comment (cesaragv) [reply hedera-node/src/test/java/com/hedera/services/state/expiry/ExpiryManagerTest.java]: Fixed

File: hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryManager.java (hunk 5 above)
Comment (qnswirlds) [reply hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryManager.java:62]: Could not mock this entity. Added some more validations after restart tracking from accounts.

File: hedera-node/data/config/bootstrap.properties [hunk 6]
```diff
@@ -1,2 +1,3 @@
 bootstrap.ledger.nodeAccounts.initialBalance=100000000000000
 bootstrap.ledger.systemAccounts.initialBalance=100000000000000
+ledger.keepRecordsInState=true
```
Comment (tinker-michaelj) [inline hedera-node/data/config/bootstrap.properties:3]: Per Nosh's comment in scrum, it seems we don't want this to override to be the default in `master`.

**Review by tinker-michaelj — APPROVED** [review]
Other than the `ledger.keepRecordsInState=true` override, LGTM!

File: hedera-node/data/config/bootstrap.properties (hunk 6 above)
Comment (qnswirlds) [reply hedera-node/data/config/bootstrap.properties]: Done removing it. Thanks!

**Review by qnswirlds — APPROVED** [review]
Discussed with @tinker-michaelj, will leave the newly added `shouldInitFees` test disabled for now because it causes PropertyLoaderTest.testAPPChangeProperties and another test in ServicesStateTest to fail with a null pointer. Test coupling will be resolved in another PR.

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

**Pull request merged**